### PR TITLE
refactor: Make InLvl a handler instead of a command

### PR DESCRIPTION
**Pull request by croyzor**

### CODE REVIEW

File: brat/Brat/Checker/Monad.hs [hunk 1]
```diff
@@ -60,8 +60,8 @@ data Context = Ctx { globalVEnv :: VEnv
 
 data CheckingSig ty where
   Fresh   :: String -> CheckingSig Name
-  -- Run a sub-process on a new namespace-level
-  InLvl   :: String -> Checking a -> CheckingSig a
+  GetNS   :: CheckingSig Namespace
```
Comment (acl-cqc) [inline brat/Brat/Checker/Monad.hs:63]: How about a single `SplitNS :: String -> CheckingSig Namespace` which both updates the namespace inside `handler` (to the second half of the split) and returns the first half of the split?

File: brat/Brat/Checker/Monad.hs [hunk 2]
```diff
@@ -213,20 +206,17 @@ handler :: Free CheckingSig v
         -> Context
         -> Graph
         -> Namespace
-        -> Either Error (v,Context,([TypedHole],Graph),Namespace)
-handler (Ret v) ctx g ns = return (v, ctx, ([], g), ns)
+        -> Either Error (v,Context,([TypedHole],Graph))
+handler (Ret v) ctx g _ = return (v, ctx, ([], g))
```
Comment (acl-cqc) [inline brat/Brat/Checker/Monad.hs:210]: I am even wondering if we could *remove* the `Namespace` parameter to handler. Have the top-level command(s) (i.e. `GetNS`+`SetNS` or `SplitNS`) die with an error, and rely on running a `localNS` inside it. You could make `run` still take a Namespace and invoke `handler (localNS c)`.... (so, a bit like `localFC`)

File: brat/Brat/Checker/Monad.hs [hunk 3]
```diff
@@ -300,7 +290,10 @@ typeErr = err . TypeErr
 
 instance FreshMonad Checking where
   freshName x = req $ Fresh x
-  str -! c = req $ InLvl str c
+  str -! c = do
```
Comment (acl-cqc) [inline brat/Brat/Checker/Monad.hs:293]: Oh, nice - I wasn't expecting it to be so easy to keep `FreshMonad`

File: brat/Brat/Load.hs [hunk 4]
```diff
@@ -153,7 +154,7 @@ loadStmtsWithEnv ns (venv, oldDecls, oldEndData) (fname, pre, stmts, cts) = addS
   let vdecls = map fst entries
   -- Now generate environment mapping usernames to nodes in the graph
   venv <- pure $ venv <> M.fromList [(name, overs) | ((name, _), (_, overs)) <- entries]
-  ((), (holes, newEndData, graph, _)) <- run venv kcStore ns $ withAliases aliases $ do
+  ((), (holes, newEndData, graph)) <- run venv kcStore newRoot $ withAliases aliases $ do
```
Comment (acl-cqc) [inline brat/Brat/Load.hs:157]: It'd be more consistent with `"globals"` to do the split outside `run` here, or to use `-!` above (as the old code did), but I'm not too bothered if you want to be different!

**Review by acl-cqc — COMMENTED** [review]
This is fab - I love that all the awkward `Req InLvl` cases are gone, and generally looks great. What do you think about `SplitNS` and/or requiring `localNS`?

File: brat/Brat/Checker/Monad.hs (hunk 2 above)
Comment (croyzor) [reply brat/Brat/Checker/Monad.hs:210]: I was thinking the same thing!

File: brat/Brat/Checker/Monad.hs (hunk 1 above)
Comment (croyzor) [reply brat/Brat/Checker/Monad.hs:63]: This makes things much nicer! 😄

File: brat/Brat/Load.hs (hunk 4 above)
Comment (croyzor) [reply brat/Brat/Load.hs:157]: I don't think we need to change the split here, since we don't do anything with the new namespace

**Review by acl-cqc — APPROVED** [review]
Super!! Thank you Craig :). I'll merge into `inference-wip/fork` when this has gone in.

**Pull request merged**